Here, making classification model for the Biclassed dataset. 


In [1]:
import pandas as pd
import seaborn as sns
import numpy as np

In [2]:
bank_data = pd.read_csv('bank-full.csv')

In [3]:
bank_data.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

In [4]:
x = bank_data[bank_data.columns[:-1]]
y = bank_data[bank_data.columns[-1]]

In [5]:
x.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome  
0  unknown    5   may       261         1     -1         0  unknown  
1  unknown    5   may       151         1     -1         0  unknown  
2  unknown    5   may        76         1     -1         0  unknown  
3  unknown    5   may        92         1     -1         0  unknown  
4  unknown    5   may       198         1     -1         0  unknown

In [6]:
#sns.pairplot(bank_data, hue = 'y')

In [7]:
y.head()

0    no
1    no
2    no
3    no
4    no
Name: y, dtype: object

Here, i wrote a function which will split data into two seperate dataframes. One is training dataframe and testing dataframe. On the basis of test split percentage.

In [8]:
def train_test_split(x, y, test_split = 0.5, rand_state = 10):
    mask = np.random.rand(x.shape[0]) < test_split
    x_test = x[mask]
    y_test = y[mask]
    x_train = x[~mask]
    y_train = y[~mask]
    return x_train, x_test, y_train, y_test

In [9]:
x_train, x_test, y_train, y_test = train_test_split(x, y, 0.2)

In [10]:
def train_test_split(x, test_split = 0.5, rand_state = 10):
    mask = np.random.rand(x.shape[0]) < test_split
    x_test = x[mask]
    x_train = x[~mask]
    return x_train, x_test

In [11]:
x_train, x_test = train_test_split(bank_data, 0.2)

In [12]:
for column in x.columns.values:
    print ("Column {}".format(column))
    print (x[column].unique())

Column age
[58 44 33 47 35 28 42 43 41 29 53 57 51 45 60 56 32 25 40 39 52 46 36 49
 59 37 50 54 55 48 24 38 31 30 27 34 23 26 61 22 21 20 66 62 83 75 67 70
 65 68 64 69 72 71 19 76 85 63 90 82 73 74 78 80 94 79 77 86 95 81 18 89
 84 87 92 93 88]
Column job
['management' 'technician' 'entrepreneur' 'blue-collar' 'unknown'
 'retired' 'admin.' 'services' 'self-employed' 'unemployed' 'housemaid'
 'student']
Column marital
['married' 'single' 'divorced']
Column education
['tertiary' 'secondary' 'unknown' 'primary']
Column default
['no' 'yes']
Column balance
[ 2143    29     2 ...  8205 14204 16353]
Column housing
['yes' 'no']
Column loan
['no' 'yes']
Column contact
['unknown' 'cellular' 'telephone']
Column day
[ 5  6  7  8  9 12 13 14 15 16 19 20 21 23 26 27 28 29 30  2  3  4 11 17
 18 24 25  1 10 22 31]
Column month
['may' 'jun' 'jul' 'aug' 'oct' 'nov' 'dec' 'jan' 'feb' 'mar' 'apr' 'sep']
Column duration
[ 261  151   76 ... 1298 1246 1556]
Column campaign
[ 1  2  3  5  4  6  7  8  9 10 11

Naive Bayes Classification is a classification algorithm. Where we are using probability of multiple features to predict the class of any probability. 
Naive bayes classifier is based on Naive Bayes Theorem. That theorem is a conditional probability which is gives the probability of any value, Where one even is already happen.

Posterior probability is the probability which comes after performing the operation or we can say posterior probability is the resultant probability. 

Prior probability is the intial probability for a every particular values. 

So,  Formula is:
$$P(A|B) = \frac{P(B|A) P(A)}{P(B)}$$
Where, <br>
$P(A|B)$ is the posterior probability. <br>
$P(A)$ is the prior probability of the class A. <br>
$P(B)$ is the prior probability of the class B. <br>


In our case, we have multiple features or we can say we have multiple values of B.
So, in this case, we always assume that all the features are independent.<br>
The formula is going to modify as.

$$P(class = Y| features...) = \frac{P(features ...| class = Y) P(class = Y)}{P(features...|class = Y)P(class = Y) + P(features... | class N) P(class = N)}$$
<br><br>
$$P\big(class = Y| age = 40\cap job = admin\cap matiral = married \cap education = secondary \cap default = yes \cap balance = 2000 \cap day = 4\cap month = jun\cap duration = 261\cap campaign = 4 \cap pday = 151\cap previous = 5\cap poutcome = other\big) = \frac{P\big(age = 40\cap job = admin\cap matiral = married \cap education = secondary \cap default = yes \cap balance = 2000 \cap day = 4\cap month = jun\cap duration = 261\cap campaign = 4 \cap pday = 151\cap previous = 5\cap poutcome = other | class = Y\big) P(class = Y)}{P\big(age = 40\cap job = admin\cap matiral = married \cap education = secondary \cap default = yes \cap balance = 2000 \cap day = 4\cap month = jun\cap duration = 261\cap campaign = 4 \cap pday = 151\cap previous = 5\cap poutcome = other | class = Y\big) P(class = Y) + P\big(age = 40\cap job = admin\cap matiral = married \cap education = secondary \cap default = yes \cap balance = 2000 \cap day = 4\cap month = jun\cap duration = 261\cap campaign = 4 \cap pday = 151\cap previous = 5\cap poutcome = other | class = N\big) P(class = N)}$$

taking some features 

$$P(class = Y | age = 40\cap job = admin) = \frac{P(age = 40 | class = Y) P(job = admin|class = Y) P(class = Y)}{P(age = 40 | class = Y) P(job = admin|class = Y) P(class = Y) + P(age = 40 | class = N) P(job = admin|class = N) P(class = N)}$$

In [13]:
class NaiveBayesClassifier (object):
    
    def __init__(self, data, target):
        
        self.data = data
        self.target = target
        self.weights = {}
        
    def estimate_mean_std(self, a, total_sample_size = 1000, each_sample_size = 10):
        
        '''
        Estimating natural paramter of normal distribution
        '''
        
        ele_list = np.array(a)
        random_sample = np.random.choice(ele_list, size = (total_sample_size, each_sample_size))
        
        ## Estimating the value of Mean
        estimate_mean = np.mean(random_sample, axis = 1)
        best_estimate_mean = np.mean(estimate_mean)
        
        ## Taking Best estimate of standard deviation
        estimate_std = np.std(random_sample, axis = 1)
        best_estimate_std = np.mean(estimate_std)
        
        return {'mu':best_estimate_mean, 'sigma' :best_estimate_std}
    
    def estimate_bino_pValue (self, a, total_sample_size = 1000, each_sample_size = 10):
        
        '''
        Estimating natural parameter of binomial distribution
        '''
        uniques = np.unique(a)
        random_sample = np.random.choice(np.array(a), size = (total_sample_size, each_sample_size))
        prob_dict = {}
        for uni in uniques:
            val = []
            for sample in random_sample:
                val.append(np.count_nonzero(uni == sample) / each_sample_size)
            prob_dict[uni] = np.mean(val)
        return prob_dict
    
    def normal_pdf(self, x, mu, sigma):
        
        '''
        Proabability distribution function for Normal Distribution
        '''
        scale_parm = 1 / (np.sqrt(2 * np.pi) * sigma)
        shift = np.e ** -(((x - mu) ** 2) / (2 * sigma ** 2))
        return scale_parm * shift
    
    def calculate_class_prob(self):
        '''
        Calculating class probability for every class
        '''
        self.unique_classes = np.unique(self.data[self.target])
        self.class_prob = {}
        for class_val in self.unique_classes:
            self.class_prob[class_val] = np.count_nonzero(self.data[self.target] == class_val) / self.data.shape[0]

    def calculate_class_based_feature_prob(self, feature):
        
        '''
        Calculating probability for particular feature on the basis of resultant classes
        '''
        val = {}
        for classes in self.unique_classes:
            mask = self.data[self.target] == classes
            feature_now = feature[mask]
            val[classes] = self.estimate_bino_pValue(feature_now)
        return val
    
    def calculate_class_based_normal_feature(self, feature):
        val = {}
        for classes in self.unique_classes:
            mask = self.data[self.target] == classes
            feature_now = feature[mask]
            val[classes] = self.estimate_mean_std(feature_now)
        return val
    
    
    def predict(self, feature):
        
        pred = {}
        pred_class_sum = 0
        pred_class = None
        total = 0
        for classes in self.unique_classes:
            pred[classes] = {}
            sumVal = 1
            # Each feature probability......
            for key, val in feature.items():
                if key in self.weights:
                    
                    if 'mu' in self.weights[key][classes].keys():
                        pred[classes][key] = self.normal_pdf(val, self.weights[key][classes]['mu'], self.weights[key][classes]['sigma'])
                        sumVal *= pred[classes][key]
                    else:
                        pred[classes][key] = self.weights[key][classes][val]
                        sumVal *= pred[classes][key]
            pred[classes]['sum'] = sumVal
            total += sumVal
        
        posterior_prob = {}
        posterior = 0
        classes = 0
        for i in self.unique_classes:
            posterior_prob[i] = pred[i]['sum'] * self.class_prob[i] / total
            if (posterior_prob[i] > posterior):
                posterior = posterior_prob[i]
                classes = i
        return classes
    
    
    def fit(self):
        
        '''
        Fitting data for every feature
        '''
        self.calculate_class_prob()
        self.weights['age'] = self.calculate_class_based_normal_feature(self.data['age'])
        self.weights['job'] = self.calculate_class_based_feature_prob(self.data['job'])
        self.weights['marital'] = self.calculate_class_based_feature_prob(self.data['marital'])
        self.weights['default'] = self.calculate_class_based_feature_prob(self.data['default'])
        self.weights['balance'] = self.calculate_class_based_normal_feature(self.data['balance'])
        self.weights['housing'] = self.calculate_class_based_feature_prob(self.data['housing'])
        self.weights['loan'] = self.calculate_class_based_feature_prob(self.data['loan'])
        self.weights['contact'] = self.calculate_class_based_feature_prob(self.data['contact'])
        self.weights['day'] = self.calculate_class_based_feature_prob(self.data['day'])
        self.weights['month'] = self.calculate_class_based_feature_prob(self.data['month'])
        self.weights['duration'] = self.calculate_class_based_normal_feature(self.data['duration'])
        self.weights['campaign'] = self.calculate_class_based_normal_feature(self.data['campaign'])
        self.weights['pdays'] = self.calculate_class_based_normal_feature(self.data['pdays'])
        self.weights['previous'] = self.calculate_class_based_normal_feature(self.data['previous'])
        self.weights['poutcome'] = self.calculate_class_based_feature_prob(self.data['poutcome'])
        
    def predict_batch(self, data):
        
        classes = data.apply(self.predict,axis = 1)
        return classes
    

In [14]:
bank_data.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y'],
      dtype='object')

In [15]:
nb = NaiveBayesClassifier(x_train, 'y')


In [16]:
nb.calculate_class_prob()
nb.class_prob

{'no': 0.8834699566669427, 'yes': 0.11653004333305733}

In [17]:
nb.fit()

In [18]:
nb.weights

{'age': {'no': {'mu': 40.966899999999995, 'sigma': 9.498752769152354},
  'yes': {'mu': 41.8992, 'sigma': 12.358790495853594}},
 'job': {'no': {'admin.': 0.11280000000000001,
   'blue-collar': 0.22290000000000004,
   'entrepreneur': 0.0305,
   'housemaid': 0.030299999999999997,
   'management': 0.20570000000000002,
   'retired': 0.0441,
   'self-employed': 0.0344,
   'services': 0.0937,
   'student': 0.0183,
   'technician': 0.17450000000000002,
   'unemployed': 0.0263,
   'unknown': 0.0065},
  'yes': {'admin.': 0.12060000000000003,
   'blue-collar': 0.1349,
   'entrepreneur': 0.024900000000000002,
   'housemaid': 0.0192,
   'management': 0.245,
   'retired': 0.0936,
   'self-employed': 0.0376,
   'services': 0.07270000000000001,
   'student': 0.0499,
   'technician': 0.1559,
   'unemployed': 0.0368,
   'unknown': 0.0089}},
 'marital': {'no': {'divorced': 0.11450000000000002,
   'married': 0.6131000000000001,
   'single': 0.27240000000000003},
  'yes': {'divorced': 0.1194, 'married': 0.

In [19]:
nb.weights['age']

{'no': {'mu': 40.966899999999995, 'sigma': 9.498752769152354},
 'yes': {'mu': 41.8992, 'sigma': 12.358790495853594}}

In [20]:
nb.predict(x_test.iloc[0].to_dict())

'no'

In [21]:
pred_classes = nb.predict_batch(x_test)

C:\Users\ritik\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:108: RuntimeWarning: invalid value encountered in double_scalars


In [22]:
class Score_Matrix(object):
    
    '''
    This class has all score matrix by which we can check efficiency of our algorithm on given dataset.
    '''
    
    def __init__(self, y_actual, y_pred):
        self.y_actual = y_actual
        self.y_pred = np.array(y_pred)
        
    def confusion_matrix(self):
        self.data = pd.crosstab(self.y_actual, self.y_pred, rownames= ['actual'], colnames = ['predicted'], margins = True)
        return self.data
        
    def recall(self):
        '''
        data:
            Dataframe of confusion matrix
        '''
        sumVal = 0
        self.confusion_matrix()
        for i in self.data.columns[:-1]:
            sumVal += (self.data[i][i] / self.data['All'][i])
        return np.round(sumVal / (self.data.shape[0] - 1), 3)
    
    def check_accuracy(self):
        sumVal = 0
        self.confusion_matrix()
        for i in self.data.columns[:-1]:
            sumVal += self.data[i][i]
        return np.round(sumVal / self.data['All']['All'], 3)
    
    def precision(self):
        sumVal = 0
        self.confusion_matrix()
        for i in self.data.columns[:-1]:
            sumVal += (self.data[i][i] / self.data[i]['All'])
        avg = sumVal / (self.data.shape[0]- 1)
        return np.round(avg,3)
    

In [23]:
sm = Score_Matrix(x_test['y'], pred_classes)

In [24]:
print ('Accuracy on test data: {}'.format(sm.check_accuracy()))

Accuracy on test data: 0.827


In [25]:
print ('Recall value on test data: {}'.format(sm.recall()))

Recall value on test data: 0.727
